File: app/src/main/java/com/example/woodworkerscompanion/data/models/BoardFootModels.kt
package com.example.woodworkerscompanion.data.models

import java.util.UUID

// MARK: - Measurement Unit
enum class MeasurementUnit(val displayName: String) {
    IMPERIAL("Imperial"),
    METRIC("Metric")
}

// MARK: - Length Unit (for Imperial)
enum class LengthUnit(val displayName: String, val symbol: String) {
    FEET("ft", "'"),
    INCHES("in", "\"")
}

// MARK: - Pricing Type
enum class PricingType(val displayName: String) {
    PER_BOARD_FOOT("Per Board Foot"),
    LINEAR("Linear")
}

// MARK: - Wood Species List
object WoodSpecies {
    val commonHardwoods = listOf(
        "Ash",
        "Black Limba",
        "Bloodwood",
        "Cherry",
        "Douglas Fir",
        "Hickory",
        "Maple",
        "Maple (Ambrosia)",
        "Maple (Birdseye)",
        "Maple (Curly)",
        "Oak (Red)",
        "Oak (White)",
        "Padauk",
        "Pine",
        "Poplar",
        "Purple Heart",
        "Tigerwood",
        "Walnut (Black)",
        "Walnut (Peruvian)",
        "Wenge",
        "Zebrawood"
    )
}

// MARK: - Board Entry
data class BoardEntry(
    val id: String = UUID.randomUUID().toString(),
    val thickness: Double? = null,
    val width: Double? = null,
    val length: Double,
    val quantity: Int,
    val unit: MeasurementUnit,
    val lengthUnit: LengthUnit? = null,  // Only used for Imperial
    val price: Double? = null,
    val pricingType: PricingType,
    val woodSpecies: String? = null
) {
    // Calculate board feet for this entry
    val boardFeet: Double
        get() {
            // Linear pricing doesn't use board feet
            if (pricingType == PricingType.LINEAR) {
                return 0.0
            }

            val t = thickness ?: return 0.0
            val w = width ?: return 0.0

            return when (unit) {
                MeasurementUnit.IMPERIAL -> {
                    // For Imperial, thickness is in quarters, so convert to inches first
                    // Board Feet = (Thickness in inches × Width in inches × Length in feet) / 12 × Quantity
                    val thicknessInInches = t / 4.0

                    // Convert length to feet if it's in inches
                    val lengthInFeet: Double = if (lengthUnit == LengthUnit.INCHES) {
                        length / 12.0
                    } else {
                        length
                    }

                    (thicknessInInches * w * lengthInFeet / 12.0) * quantity
                }
                MeasurementUnit.METRIC -> {
                    // Convert to board feet: 1 board foot = 2359.737 cm³
                    // Thickness (cm) × Width (cm) × Length (cm) / 2359.737 × Quantity
                    (t * w * length / 2359.737) * quantity
                }
            }
        }

    // Calculate cost for this entry
    val cost: Double
        get() {
            val p = price ?: return 0.0

            return when (pricingType) {
                PricingType.PER_BOARD_FOOT -> boardFeet * p
                PricingType.LINEAR -> {
                    // For linear, just multiply length by price and quantity
                    length * p * quantity
                }
            }
        }

    // Formatted display
    val displayString: String
        get() {
            val quantityStr = if (quantity > 1) "$quantity × " else ""
            val speciesStr = woodSpecies?.let { " - $it" } ?: ""

            if (pricingType == PricingType.LINEAR) {
                // Linear only shows length
                return when (unit) {
                    MeasurementUnit.IMPERIAL -> {
                        val lengthSymbol = if (lengthUnit == LengthUnit.INCHES) "\"" else "'"
                        "$quantityStr$length$lengthSymbol$speciesStr"
                    }
                    MeasurementUnit.METRIC -> {
                        "${quantityStr}${length}cm$speciesStr"
                    }
                }
            }

            val t = thickness ?: return ""
            val w = width ?: return ""

            return when (unit) {
                MeasurementUnit.IMPERIAL -> {
                    // Show thickness in quarters format
                    val thicknessInt = t.toInt()
                    val lengthSymbol = if (lengthUnit == LengthUnit.INCHES) "\"" else "'"
                    "$quantityStr$thicknessInt/4\" × $w\" × $length$lengthSymbol$speciesStr"
                }
                MeasurementUnit.METRIC -> {
                    "${quantityStr}${t}cm × ${w}cm × ${length}cm$speciesStr"
                }
            }
        }

    // Formatted display without species (for export text)
    val displayStringWithoutSpecies: String
        get() {
            val quantityStr = if (quantity > 1) "$quantity × " else ""

            if (pricingType == PricingType.LINEAR) {
                // Linear only shows length
                return when (unit) {
                    MeasurementUnit.IMPERIAL -> {
                        val lengthSymbol = if (lengthUnit == LengthUnit.INCHES) "\"" else "'"
                        "$quantityStr$length$lengthSymbol"
                    }
                    MeasurementUnit.METRIC -> {
                        "${quantityStr}${length}cm"
                    }
                }
            }

            val t = thickness ?: return ""
            val w = width ?: return ""

            return when (unit) {
                MeasurementUnit.IMPERIAL -> {
                    // Show thickness in quarters format
                    val thicknessInt = t.toInt()
                    val lengthSymbol = if (lengthUnit == LengthUnit.INCHES) "\"" else "'"
                    "$quantityStr$thicknessInt/4\" × $w\" × $length$lengthSymbol"
                }
                MeasurementUnit.METRIC -> {
                    "${quantityStr}${t}cm × ${w}cm × ${length}cm"
                }
            }
        }
}

// MARK: - Common Lumber Sizes
data class LumberPreset(
    val name: String,
    val thickness: Double,
    val width: Double
) {
    companion object {
        val imperialPresets = listOf(
            LumberPreset("1×2", 0.75, 1.5),
            LumberPreset("1×3", 0.75, 2.5),
            LumberPreset("1×4", 0.75, 3.5),
            LumberPreset("1×6", 0.75, 5.5),
            LumberPreset("1×8", 0.75, 7.25),
            LumberPreset("1×10", 0.75, 9.25),
            LumberPreset("1×12", 0.75, 11.25),
            LumberPreset("2×4", 1.5, 3.5),
            LumberPreset("2×6", 1.5, 5.5),
            LumberPreset("2×8", 1.5, 7.25),
            LumberPreset("2×10", 1.5, 9.25),
            LumberPreset("2×12", 1.5, 11.25),
            LumberPreset("4×4", 3.5, 3.5),
            LumberPreset("6×6", 5.5, 5.5)
        )

        val metricPresets = listOf(
            LumberPreset("2×5", 2.0, 5.0),
            LumberPreset("2×10", 2.0, 10.0),
            LumberPreset("2×15", 2.0, 15.0),
            LumberPreset("3×10", 3.0, 10.0),
            LumberPreset("3×15", 3.0, 15.0),
            LumberPreset("4×10", 4.0, 10.0),
            LumberPreset("4×15", 4.0, 15.0),
            LumberPreset("5×10", 5.0, 10.0),
            LumberPreset("5×15", 5.0, 15.0),
            LumberPreset("5×20", 5.0, 20.0)
        )
    }
}

File: iOS/BoardFootCalculator/BoardFootModels.swift
//
//  BoardFootModels.swift
//  Woodworker's Companion
//
//  Created by James Corey on 10/15/25.
//

import Foundation

// MARK: - Measurement Unit
enum MeasurementUnit: String, CaseIterable, Codable {
  case imperial = "Imperial"
  case metric = "Metric"
}

// MARK: - Length Unit (for Imperial)
enum LengthUnit: String, CaseIterable, Codable {
  case feet = "ft"
  case inches = "in"
}

// MARK: - Pricing Type
enum PricingType: String, CaseIterable, Codable {
  case perBoardFoot = "Per Board Foot"
  case linear = "Linear"
}

// MARK: - Wood Species List
struct WoodSpecies {
  static let commonHardwoods: [String] = [
    "Ash",
    "Black Limba",
    "Bloodwood",
    "Cherry",
    "Douglas Fir",
    "Hickory",
    "Maple",
    "Maple (Ambrosia)",
    "Maple (Birdseye)",
    "Maple (Curly)",
    "Oak (Red)",
    "Oak (White)",
    "Padauk",
    "Pine",
    "Poplar",
    "Purple Heart",
    "Tigerwood",
    "Walnut (Black)",
    "Walnut (Peruvian)",
    "Wenge",
    "Zebrawood",
  ]
}

// MARK: - Board Entry
struct BoardEntry: Identifiable, Codable, Equatable {
  let id: UUID
  let thickness: Double?
  let width: Double?
  let length: Double
  let quantity: Int
  let unit: MeasurementUnit
  let lengthUnit: LengthUnit?  // Only used for Imperial
  let price: Double?
  let pricingType: PricingType
  let woodSpecies: String?

  init(
    id: UUID = UUID(), thickness: Double?, width: Double?, length: Double, quantity: Int,
    unit: MeasurementUnit, lengthUnit: LengthUnit?, price: Double?, pricingType: PricingType,
    woodSpecies: String?
  ) {
    self.id = id
    self.thickness = thickness
    self.width = width
    self.length = length
    self.quantity = quantity
    self.unit = unit
    self.lengthUnit = lengthUnit
    self.price = price
    self.pricingType = pricingType
    self.woodSpecies = woodSpecies
  }

  // Equatable conformance
  static func == (lhs: BoardEntry, rhs: BoardEntry) -> Bool {
    lhs.id == rhs.id
  }

  // Calculate board feet for this entry
  var boardFeet: Double {
    // Linear pricing doesn't use board feet
    if pricingType == .linear {
      return 0
    }

    guard let thickness = thickness, let width = width else { return 0 }

    switch unit {
    case .imperial:
      // For Imperial, thickness is in quarters, so convert to inches first
      // Board Feet = (Thickness in inches × Width in inches × Length in feet) / 12 × Quantity
      let thicknessInInches = thickness / 4.0

      // Convert length to feet if it's in inches
      let lengthInFeet: Double
      if lengthUnit == .inches {
        lengthInFeet = length / 12.0
      } else {
        lengthInFeet = length
      }

      return (thicknessInInches * width * lengthInFeet / 12.0) * Double(quantity)
    case .metric:
      // Convert to board feet: 1 board foot = 2359.737 cm³
      // Thickness (cm) × Width (cm) × Length (cm) / 2359.737 × Quantity
      return (thickness * width * length / 2359.737) * Double(quantity)
    }
  }

  // Calculate cost for this entry
  var cost: Double {
    guard let price = price else { return 0 }

    switch pricingType {
    case .perBoardFoot:
      return boardFeet * price
    case .linear:
      // For linear, just multiply length by price and quantity
      return length * price * Double(quantity)
    }
  }

  // Formatted display
  var displayString: String {
    let quantityStr = quantity > 1 ? "\(quantity) × " : ""
    let speciesStr = woodSpecies.map { " - \($0)" } ?? ""

    if pricingType == .linear {
      // Linear only shows length
      switch unit {
      case .imperial:
        let lengthSymbol = lengthUnit == .inches ? "\"" : "'"
        return "\(quantityStr)\(length)\(lengthSymbol)\(speciesStr)"
      case .metric:
        return "\(quantityStr)\(length)cm\(speciesStr)"
      }
    }

    guard let thickness = thickness, let width = width else { return "" }

    switch unit {
    case .imperial:
      // Show thickness in quarters format
      let thicknessInt = Int(thickness)
      let lengthSymbol = lengthUnit == .inches ? "\"" : "'"
      return
        "\(quantityStr)\(thicknessInt)/4\" × \(width)\" × \(length)\(lengthSymbol)\(speciesStr)"
    case .metric:
      return "\(quantityStr)\(thickness)cm × \(width)cm × \(length)cm\(speciesStr)"
    }
  }

  // Formatted display without species (for export text)
  var displayStringWithoutSpecies: String {
    let quantityStr = quantity > 1 ? "\(quantity) × " : ""

    if pricingType == .linear {
      // Linear only shows length
      switch unit {
      case .imperial:
        let lengthSymbol = lengthUnit == .inches ? "\"" : "'"
        return "\(quantityStr)\(length)\(lengthSymbol)"
      case .metric:
        return "\(quantityStr)\(length)cm"
      }
    }

    guard let thickness = thickness, let width = width else { return "" }

    switch unit {
    case .imperial:
      // Show thickness in quarters format
      let thicknessInt = Int(thickness)
      let lengthSymbol = lengthUnit == .inches ? "\"" : "'"
      return
        "\(quantityStr)\(thicknessInt)/4\" × \(width)\" × \(length)\(lengthSymbol)"
    case .metric:
      return "\(quantityStr)\(thickness)cm × \(width)cm × \(length)cm"
    }
  }
}

// MARK: - Common Lumber Sizes
struct LumberPreset {
  let name: String
  let thickness: Double
  let width: Double

  static let imperialPresets: [LumberPreset] = [
    LumberPreset(name: "1×2", thickness: 0.75, width: 1.5),
    LumberPreset(name: "1×3", thickness: 0.75, width: 2.5),
    LumberPreset(name: "1×4", thickness: 0.75, width: 3.5),
    LumberPreset(name: "1×6", thickness: 0.75, width: 5.5),
    LumberPreset(name: "1×8", thickness: 0.75, width: 7.25),
    LumberPreset(name: "1×10", thickness: 0.75, width: 9.25),
    LumberPreset(name: "1×12", thickness: 0.75, width: 11.25),
    LumberPreset(name: "2×4", thickness: 1.5, width: 3.5),
    LumberPreset(name: "2×6", thickness: 1.5, width: 5.5),
    LumberPreset(name: "2×8", thickness: 1.5, width: 7.25),
    LumberPreset(name: "2×10", thickness: 1.5, width: 9.25),
    LumberPreset(name: "2×12", thickness: 1.5, width: 11.25),
    LumberPreset(name: "4×4", thickness: 3.5, width: 3.5),
    LumberPreset(name: "6×6", thickness: 5.5, width: 5.5),
  ]

  static let metricPresets: [LumberPreset] = [
    LumberPreset(name: "2×5", thickness: 2, width: 5),
    LumberPreset(name: "2×10", thickness: 2, width: 10),
    LumberPreset(name: "2×15", thickness: 2, width: 15),
    LumberPreset(name: "3×10", thickness: 3, width: 10),
    LumberPreset(name: "3×15", thickness: 3, width: 15),
    LumberPreset(name: "4×10", thickness: 4, width: 10),
    LumberPreset(name: "4×15", thickness: 4, width: 15),
    LumberPreset(name: "5×10", thickness: 5, width: 10),
    LumberPreset(name: "5×15", thickness: 5, width: 15),
    LumberPreset(name: "5×20", thickness: 5, width: 20),
  ]
}
